crates/bitwarden/src/platform/fido2/authenticator.rs

```diff
@@ -92,242 +90,373 @@ impl<'a> Fido2Authenticator<'a> {
+                .await?;
+
+            let cred = Fido2CredentialFullView::try_from_credential(cred, user, rp)?;
+
+            let mut picked = this
+                .authenticator
+                .user_interface
+                .pick_credential_for_creation(creds, cred.clone().into())
+                .await?;
+
+            let enc = this.authenticator.client.get_encryption_settings()?;
+
+            picked.set_new_fido2_credentials(enc, vec![cred])?;
+
+            // Store the selected credential for later use
+            this.authenticator
+                .selected_credential
+                .lock()
+                .expect("Mutex is not poisoned")
+                .replace(picked.clone());
+
+            let key = enc
+                .get_key(&picked.organization_id)
+                .ok_or(Error::VaultLocked)?;
+            let encrypted = picked.encrypt_with_key(key)?;
+
+            this.authenticator
+                .credential_store
+                .save_credential(encrypted)
+                .await?;
+
+            Ok(())
+        }
+
+        inner(self, cred, user, rp).await.map_err(|e| {
+            error!("Error saving credential: {e:?}");
+            VendorError::try_from(0xF1)
+                .expect("Valid vendor error code")
+                .into()
+        })
     }
 
     async fn update_credential(&mut self, cred: Passkey) -> Result<(), StatusCode> {
-        Ok(())
+        // This is just a wrapper around the actual implementation to allow for ? error handling
+        async fn inner(this: &mut CredentialStoreImpl<'_>, cred: Passkey) -> Result<()> {
+            let mut creds = this
+                .authenticator
+                .credential_store
+                .find_credentials(None, cred.rp_id.clone())
+                .await?;
+
+            // Get the credential with the same id as the one we're updating
+            creds.retain(|c| {
+                c.login
+                    .as_ref()
+                    .and_then(|l| l.fido2_credentials.as_ref())
+                    .and_then(|f| f.first())
+                    .map(|cipher_cred| {
+                        cipher_cred.credential_id.expose().as_bytes() == cred.credential_id.deref()
+                    })
+                    .unwrap_or(false)
+            });
+            let cred = match creds.len() {
+                1 => creds.into_iter().next().expect("Vec has one element"),
+                _ => return Err("Only one credential should match the id".into()),
+            };
+
+            let enc = this.authenticator.client.get_encryption_settings()?;
+
+            // Store the selected credential for later use
+            this.authenticator
+                .selected_credential
+                .lock()
+                .expect("Mutex is not poisoned")
+                .replace(cred.clone());
+
+            let key = enc
+                .get_key(&cred.organization_id)
+                .ok_or(Error::VaultLocked)?;
```

Comment:
thought: this should be fine since `find_credentials` should unlock the vault

Reply:
Yeah I think this is like the `require!` case above, where using `expect` would yield the same result as the operation should never fail.

As to why using Result instead of `.expect()`, the function already returned a `Result` type so that seemed like the easiest path.